Array of strings incorrectly indexed as `dense_vector`

### Elasticsearch Version

8.11+

### Installed Plugins

_No response_

### Java Version

_bundled_

### OS Version

ANY

### Problem Description

When indexing an array of string values, these may be dynamically indexed as `dense_vector`. This is unexpected behavior as they should be indexed as `text`.

The error the user will see is: `:"org.elasticsearch.common.ParsingException: Failed to parse object: expecting token of type [VALUE_NUMBER] but found [VALUE_STRING]`

### Steps to Reproduce

The following yaml test reproduces the bug:
```
---
"Fields indexed as strings won't be transformed into dense_vector":
  - do:
      index:
        index: strings-are-not-floats
        refresh: true
        body:
          obviously_string: ["foo", "foo", "foo", "foo", "foo", "foo", "foo", "foo", "foo", "foo",
                           "foo", "foo", "foo", "foo", "foo", "foo", "foo", "foo", "foo", "foo",
                           "foo", "foo", "foo", "foo", "foo", "foo", "foo", "foo", "foo", "foo",
                           "foo", "foo", "foo", "foo", "foo", "foo", "foo", "foo", "foo", "foo",
                           "foo", "foo", "foo", "foo", "foo", "foo", "foo", "foo", "foo", "foo",
                           "foo", "foo", "foo", "foo", "foo", "foo", "foo", "foo", "foo", "foo",
                           "foo", "foo", "foo", "foo", "foo", "foo", "foo", "foo", "foo", "foo",
                           "foo", "foo", "foo", "foo", "foo", "foo", "foo", "foo", "foo", "foo",
                           "foo", "foo", "foo", "foo", "foo", "foo", "foo", "foo", "foo", "foo",
                           "foo", "foo", "foo", "foo", "foo", "foo", "foo", "foo", "foo", "foo",
                           "foo", "foo", "foo", "foo", "foo", "foo", "foo", "foo", "foo", "foo",
                           "foo", "foo", "foo", "foo", "foo", "foo", "foo", "foo", "foo", "foo",
                           "foo", "foo", "foo", "foo", "foo", "foo", "foo", "foo"]
  - do:
      cluster.health:
        wait_for_events: languid
  - do:
      indices.get_mapping:
        index: strings-are-not-floats

  - match: { strings-are-not-floats.mappings.properties.obviously_string.type: text }
```

### Logs (if relevant)

```
:"org.elasticsearch.common.ParsingException: Failed to parse object: expecting token of type [VALUE_NUMBER] but found [VALUE_STRING]
  at org.elasticsearch.server@8.12.0-SNAPSHOT/org.elasticsearch.common.xcontent.XContentParserUtils.parsingException(XContentParserUtils.java:76)
  at org.elasticsearch.server@8.12.0-SNAPSHOT/org.elasticsearch.common.xcontent.XContentParserUtils.ensureExpectedToken(XContentParserUtils.java:69)
  at org.elasticsearch.server@8.12.0-SNAPSHOT/org.elasticsearch.index.mapper.vectors.DenseVectorFieldMapper$ElementType$2.parseKnnVector(DenseVectorFieldMapper.java:489)
  at org.elasticsearch.server@8.12.0-SNAPSHOT/org.elasticsearch.index.mapper.vectors.DenseVectorFieldMapper.parseKnnVector(DenseVectorFieldMapper.java:991)
  at org.elasticsearch.server@8.12.0-SNAPSHOT/org.elasticsearch.index.mapper.vectors.DenseVectorFieldMapper.parse(DenseVectorFieldMapper.java:986)
  at org.elasticsearch.server@8.12.0-SNAPSHOT/org.elasticsearch.index.mapper.DocumentParser.parseObjectOrField(DocumentParser.java:411)
  at org.elasticsearch.server@8.12.0-SNAPSHOT/org.elasticsearch.index.mapper.DocumentParser.parseArray(DocumentParser.java:547)
  at org.elasticsearch.server@8.12.0-SNAPSHOT/org.elasticsearch.index.mapper.DocumentParser.innerParseObject(DocumentParser.java:335)
  at org.elasticsearch.server@8.12.0-SNAPSHOT/org.elasticsearch.index.mapper.DocumentParser.parseObjectOrNested(DocumentParser.java:293)
  at org.elasticsearch.server@8.12.0-SNAPSHOT/org.elasticsearch.index.mapper.DocumentParser.internalParseDocument(DocumentParser.java:144)
  at org.elasticsearch.server@8.12.0-SNAPSHOT/org.elasticsearch.index.mapper.DocumentParser.parseDocument(DocumentParser.java:96)
  at org.elasticsearch.server@8.12.0-SNAPSHOT/org.elasticsearch.index.mapper.DocumentMapper.parse(DocumentMapper.java:92)
  at org.elasticsearch.server@8.12.0-SNAPSHOT/org.elasticsearch.index.shard.IndexShard.prepareIndex(IndexShard.java:1011)
  at org.elasticsearch.server@8.12.0-SNAPSHOT/org.elasticsearch.index.shard.IndexShard.applyIndexOperation(IndexShard.java:959)
  at org.elasticsearch.server@8.12.0-SNAPSHOT/org.elasticsearch.index.shard.IndexShard.applyIndexOperationOnPrimary(IndexShard.java:903)
  at org.elasticsearch.server@8.12.0-SNAPSHOT/org.elasticsearch.action.bulk.TransportShardBulkAction.executeBulkItemRequest(TransportShardBulkAction.java:360)
  at org.elasticsearch.server@8.12.0-SNAPSHOT/org.elasticsearch.action.bulk.TransportShardBulkAction$2.doRun(TransportShardBulkAction.java:224)
  at org.elasticsearch.server@8.12.0-SNAPSHOT/org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:26)
  at org.elasticsearch.server@8.12.0-SNAPSHOT/org.elasticsearch.common.util.concurrent.TimedRunnable.doRun(TimedRunnable.java:33)
  at org.elasticsearch.server@8.12.0-SNAPSHOT/org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingAbstractRunnable.doRun(ThreadContext.java:983)
  at org.elasticsearch.server@8.12.0-SNAPSHOT/org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:26)
  at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
  at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
  at java.base/java.lang.Thread.run(Thread.java:1583)
"}],"type":"document_parsing_exception","reason":"[-1:19] failed to parse: Failed to parse object: expecting token of type [VALUE_NUMBER] but found [VALUE_STRING]","caused_by":{"type":"parsing_exception","reason":"Failed to parse object: expecting token of type [VALUE_NUMBER] but found [VALUE_STRING]","stack_trace":"org.elasticsearch.common.ParsingException: Failed to parse object: expecting token of type [VALUE_NUMBER] but found [VALUE_STRING]
  at org.elasticsearch.server@8.12.0-SNAPSHOT/org.elasticsearch.common.xcontent.XContentParserUtils.parsingException(XContentParserUtils.java:76)
  at org.elasticsearch.server@8.12.0-SNAPSHOT/org.elasticsearch.common.xcontent.XContentParserUtils.ensureExpectedToken(XContentParserUtils.java:69)
  at org.elasticsearch.server@8.12.0-SNAPSHOT/org.elasticsearch.index.mapper.vectors.DenseVectorFieldMapper$ElementType$2.parseKnnVector(DenseVectorFieldMapper.java:489)
  at org.elasticsearch.server@8.12.0-SNAPSHOT/org.elasticsearch.index.mapper.vectors.DenseVectorFieldMapper.parseKnnVector(DenseVectorFieldMapper.java:991)
  at org.elasticsearch.server@8.12.0-SNAPSHOT/org.elasticsearch.index.mapper.vectors.DenseVectorFieldMapper.parse(DenseVectorFieldMapper.java:986)
  at org.elasticsearch.server@8.12.0-SNAPSHOT/org.elasticsearch.index.mapper.DocumentParser.parseObjectOrField(DocumentParser.java:411)
  at org.elasticsearch.server@8.12.0-SNAPSHOT/org.elasticsearch.index.mapper.DocumentParser.parseArray(DocumentParser.java:547)
  at org.elasticsearch.server@8.12.0-SNAPSHOT/org.elasticsearch.index.mapper.DocumentParser.innerParseObject(DocumentParser.java:335)
  at org.elasticsearch.server@8.12.0-SNAPSHOT/org.elasticsearch.index.mapper.DocumentParser.parseObjectOrNested(DocumentParser.java:293)
  at org.elasticsearch.server@8.12.0-SNAPSHOT/org.elasticsearch.index.mapper.DocumentParser.internalParseDocument(DocumentParser.java:144)
  at org.elasticsearch.server@8.12.0-SNAPSHOT/org.elasticsearch.index.mapper.DocumentParser.parseDocument(DocumentParser.java:96)
  at org.elasticsearch.server@8.12.0-SNAPSHOT/org.elasticsearch.index.mapper.DocumentMapper.parse(DocumentMapper.java:92)
  at org.elasticsearch.server@8.12.0-SNAPSHOT/org.elasticsearch.index.shard.IndexShard.prepareIndex(IndexShard.java:1011)
  at org.elasticsearch.server@8.12.0-SNAPSHOT/org.elasticsearch.index.shard.IndexShard.applyIndexOperation(IndexShard.java:959)
  at org.elasticsearch.server@8.12.0-SNAPSHOT/org.elasticsearch.index.shard.IndexShard.applyIndexOperationOnPrimary(IndexShard.java:903)
  at org.elasticsearch.server@8.12.0-SNAPSHOT/org.elasticsearch.action.bulk.TransportShardBulkAction.executeBulkItemRequest(TransportShardBulkAction.java:360)
  at org.elasticsearch.server@8.12.0-SNAPSHOT/org.elasticsearch.action.bulk.TransportShardBulkAction$2.doRun(TransportShardBulkAction.java:224)
  at org.elasticsearch.server@8.12.0-SNAPSHOT/org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:26)
  at org.elasticsearch.server@8.12.0-SNAPSHOT/org.elasticsearch.common.util.concurrent.TimedRunnable.doRun(TimedRunnable.java:33)
  at org.elasticsearch.server@8.12.0-SNAPSHOT/org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingAbstractRunnable.doRun(ThreadContext.java:983)
  at org.elasticsearch.server@8.12.0-SNAPSHOT/org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:26)
  at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
  at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
  at java.base/java.lang.Thread.run(Thread.java:1583)
"},"stack_trace":"org.elasticsearch.index.mapper.DocumentParsingException: [-1:19] failed to parse: Failed to parse object: expecting token of type [VALUE_NUMBER] but found [VALUE_STRING]
  at org.elasticsearch.server@8.12.0-SNAPSHOT/org.elasticsearch.index.mapper.DocumentParser.wrapInDocumentParsingException(DocumentParser.java:246)
  at org.elasticsearch.server@8.12.0-SNAPSHOT/org.elasticsearch.index.mapper.DocumentParser.internalParseDocument(DocumentParser.java:153)
  at org.elasticsearch.server@8.12.0-SNAPSHOT/org.elasticsearch.index.mapper.DocumentParser.parseDocument(DocumentParser.java:96)
  at org.elasticsearch.server@8.12.0-SNAPSHOT/org.elasticsearch.index.mapper.DocumentMapper.parse(DocumentMapper.java:92)
  at org.elasticsearch.server@8.12.0-SNAPSHOT/org.elasticsearch.index.shard.IndexShard.prepareIndex(IndexShard.java:1011)
  at org.elasticsearch.server@8.12.0-SNAPSHOT/org.elasticsearch.index.shard.IndexShard.applyIndexOperation(IndexShard.java:959)
  at org.elasticsearch.server@8.12.0-SNAPSHOT/org.elasticsearch.index.shard.IndexShard.applyIndexOperationOnPrimary(IndexShard.java:903)
  at org.elasticsearch.server@8.12.0-SNAPSHOT/org.elasticsearch.action.bulk.TransportShardBulkAction.executeBulkItemRequest(TransportShardBulkAction.java:360)
  at org.elasticsearch.server@8.12.0-SNAPSHOT/org.elasticsearch.action.bulk.TransportShardBulkAction$2.doRun(TransportShardBulkAction.java:224)
  at org.elasticsearch.server@8.12.0-SNAPSHOT/org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:26)
  at org.elasticsearch.server@8.12.0-SNAPSHOT/org.elasticsearch.common.util.concurrent.TimedRunnable.doRun(TimedRunnable.java:33)
  at org.elasticsearch.server@8.12.0-SNAPSHOT/org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingAbstractRunnable.doRun(ThreadContext.java:983)
  at org.elasticsearch.server@8.12.0-SNAPSHOT/org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:26)
  at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
  at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
  at java.base/java.lang.Thread.run(Thread.java:1583)
Caused by: org.elasticsearch.common.ParsingException: Failed to parse object: expecting token of type [VALUE_NUMBER] but found [VALUE_STRING]
  at org.elasticsearch.server@8.12.0-SNAPSHOT/org.elasticsearch.common.xcontent.XContentParserUtils.parsingException(XContentParserUtils.java:76)
  at org.elasticsearch.server@8.12.0-SNAPSHOT/org.elasticsearch.common.xcontent.XContentParserUtils.ensureExpectedToken(XContentParserUtils.java:69)
  at org.elasticsearch.server@8.12.0-SNAPSHOT/org.elasticsearch.index.mapper.vectors.DenseVectorFieldMapper$ElementType$2.parseKnnVector(DenseVectorFieldMapper.java:489)
  at org.elasticsearch.server@8.12.0-SNAPSHOT/org.elasticsearch.index.mapper.vectors.DenseVectorFieldMapper.parseKnnVector(DenseVectorFieldMapper.java:991)
  at org.elasticsearch.server@8.12.0-SNAPSHOT/org.elasticsearch.index.mapper.vectors.DenseVectorFieldMapper.parse(DenseVectorFieldMapper.java:986)
  at org.elasticsearch.server@8.12.0-SNAPSHOT/org.elasticsearch.index.mapper.DocumentParser.parseObjectOrField(DocumentParser.java:411)
  at org.elasticsearch.server@8.12.0-SNAPSHOT/org.elasticsearch.index.mapper.DocumentParser.parseArray(DocumentParser.java:547)
  at org.elasticsearch.server@8.12.0-SNAPSHOT/org.elasticsearch.index.mapper.DocumentParser.innerParseObject(DocumentParser.java:335)
  at org.elasticsearch.server@8.12.0-SNAPSHOT/org.elasticsearch.index.mapper.DocumentParser.parseObjectOrNested(DocumentParser.java:293)
  at org.elasticsearch.server@8.12.0-SNAPSHOT/org.elasticsearch.index.mapper.DocumentParser.internalParseDocument(DocumentParser.java:144)
  ... 14 more
```

--- a/docs/changelog/101967.yaml
+++ b/docs/changelog/101967.yaml
@@ -1,0 +1,5 @@
+pr: 101967
+summary: "Fix incorrect dynamic mapping for non-numeric-value arrays #101965"
+area: Mapping
+type: bug
+issues: []

--- a/rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/search.vectors/60_dense_vector_dynamic_mapping.yml
+++ b/rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/search.vectors/60_dense_vector_dynamic_mapping.yml
@@ -2,7 +2,38 @@ setup:
   - skip:
       version: ' - 8.10.99'
       reason: 'Dynamic mapping of floats to dense_vector was added in 8.11'
+---
+"Fields indexed as strings won't be transformed into dense_vector":
+  - skip:
+      # TODO adjust after backport
+      version: ' - 8.11.99'
+      reason: 'Dynamic mapping of things other than numbers is fixed in 8.12'
+  - do:
+      index:
+        index: strings-are-not-floats
+        refresh: true
+        body:
+          obviously_string: ["foo", "foo", "foo", "foo", "foo", "foo", "foo", "foo", "foo", "foo",
+                           "foo", "foo", "foo", "foo", "foo", "foo", "foo", "foo", "foo", "foo",
+                           "foo", "foo", "foo", "foo", "foo", "foo", "foo", "foo", "foo", "foo",
+                           "foo", "foo", "foo", "foo", "foo", "foo", "foo", "foo", "foo", "foo",
+                           "foo", "foo", "foo", "foo", "foo", "foo", "foo", "foo", "foo", "foo",
+                           "foo", "foo", "foo", "foo", "foo", "foo", "foo", "foo", "foo", "foo",
+                           "foo", "foo", "foo", "foo", "foo", "foo", "foo", "foo", "foo", "foo",
+                           "foo", "foo", "foo", "foo", "foo", "foo", "foo", "foo", "foo", "foo",
+                           "foo", "foo", "foo", "foo", "foo", "foo", "foo", "foo", "foo", "foo",
+                           "foo", "foo", "foo", "foo", "foo", "foo", "foo", "foo", "foo", "foo",
+                           "foo", "foo", "foo", "foo", "foo", "foo", "foo", "foo", "foo", "foo",
+                           "foo", "foo", "foo", "foo", "foo", "foo", "foo", "foo", "foo", "foo",
+                           "foo", "foo", "foo", "foo", "foo", "foo", "foo", "foo"]
+  - do:
+      cluster.health:
+        wait_for_events: languid
+  - do:
+      indices.get_mapping:
+        index: strings-are-not-floats
 
+  - match: { strings-are-not-floats.mappings.properties.obviously_string.type: text }
 ---
 "Fields with float arrays below the threshold still map as float":
 

--- a/server/src/main/java/org/elasticsearch/index/mapper/DocumentParser.java
+++ b/server/src/main/java/org/elasticsearch/index/mapper/DocumentParser.java
@@ -613,9 +613,12 @@ public final class DocumentParser {
                 || context.isCopyToField(fullFieldName)
                 || mappers.size() < MIN_DIMS_FOR_DYNAMIC_FLOAT_MAPPING
                 || mappers.size() > MAX_DIMS_COUNT
-                || mappers.stream().anyMatch(m -> m instanceof NumberFieldMapper.Builder == false)
+                // Anything that is NOT a number or anything that IS a number but not mapped to `float` should NOT be mapped to dense_vector
                 || mappers.stream()
-                    .allMatch(m -> m instanceof NumberFieldMapper.Builder nb && nb.type != NumberFieldMapper.NumberType.FLOAT)) {
+                    .anyMatch(
+                        m -> m instanceof NumberFieldMapper.Builder == false
+                            || ((NumberFieldMapper.Builder) m).type != NumberFieldMapper.NumberType.FLOAT
+                    )) {
                 return;
             }
 

--- a/server/src/main/java/org/elasticsearch/index/mapper/DocumentParser.java
+++ b/server/src/main/java/org/elasticsearch/index/mapper/DocumentParser.java
@@ -613,6 +613,7 @@ public final class DocumentParser {
                 || context.isCopyToField(fullFieldName)
                 || mappers.size() < MIN_DIMS_FOR_DYNAMIC_FLOAT_MAPPING
                 || mappers.size() > MAX_DIMS_COUNT
+                || mappers.stream().anyMatch(m -> m instanceof NumberFieldMapper.Builder == false)
                 || mappers.stream()
                     .allMatch(m -> m instanceof NumberFieldMapper.Builder nb && nb.type != NumberFieldMapper.NumberType.FLOAT)) {
                 return;

--- a/server/src/test/java/org/elasticsearch/index/mapper/DynamicMappingTests.java
+++ b/server/src/test/java/org/elasticsearch/index/mapper/DynamicMappingTests.java
@@ -21,6 +21,7 @@ import org.elasticsearch.xcontent.XContentType;
 
 import java.io.IOException;
 import java.time.Instant;
+import java.util.stream.Stream;
 
 import static org.elasticsearch.index.mapper.vectors.DenseVectorFieldMapper.MAX_DIMS_COUNT;
 import static org.elasticsearch.index.mapper.vectors.DenseVectorFieldMapper.MIN_DIMS_FOR_DYNAMIC_FLOAT_MAPPING;
@@ -990,5 +991,16 @@ public class DynamicMappingTests extends MapperServiceTestCase {
         assertNotNull(update);
         ObjectMapper parent = (ObjectMapper) update.getRoot().getMapper("parent_object");
         assertThat(((FieldMapper) parent.getMapper("mapsToDenseVector")).fieldType().typeName(), equalTo("dense_vector"));
+    }
+
+    public void testStringArraysAreText() throws IOException {
+        DocumentMapper mapper = createDocumentMapper(topMapping(b -> b.field("numeric_detection", true)));
+        BytesReference source = BytesReference.bytes(
+            XContentFactory.jsonBuilder().startObject().field("mapsToString", Stream.generate(() -> "foo").limit(129).toArray()).endObject()
+        );
+        ParsedDocument parsedDocument = mapper.parse(new SourceToParse("id", source, XContentType.JSON));
+        Mapping update = parsedDocument.dynamicMappingsUpdate();
+        assertNotNull(update);
+        assertThat(((FieldMapper) update.getRoot().getMapper("mapsToString")).fieldType().typeName(), equalTo("text"));
     }
 }

